### ISSUE TITLE
Ensure that perses calculations can be resumed

### ISSUE THREAD

**Issue by jchodera** [choderalab/perses#816]
We've received reports that the perses calculations hang when trying to restart/resume.

We should make sure that perses calculations can be resumed by
* adding unit tests to run one iteration and terminate, then resume via the API
* also test the same via the CLI
* also test early termination via `SIGKILL` during the startup phase to make sure the calculation is not left in an un-rerunnable state
* test on our batch queue system
* test on AWS

**Comment by mikemhenry** [comment]
@zhang-ivy example for small molecule transformation (protein mutation would be nice but lower priority)
@jchodera link to yank code that handles resume

**Comment by jchodera** [comment]
The YANK driver code explicitly checks to see if there is a calculation that can be resumed [here](https://github.com/choderalab/yank/blob/bb2b8657896b73950a9f86cbb2c3d5832e0fa692/Yank/experiment.py#L400-L411). That code is located in `AlchemicalPhase`, which just [defers to the `MultiStateSampler`](https://github.com/choderalab/yank/blob/bb2b8657896b73950a9f86cbb2c3d5832e0fa692/Yank/yank.py#L813-L836). If the file exists, it [resumes the `MultiStateSampler` object from storage](https://github.com/choderalab/yank/blob/bb2b8657896b73950a9f86cbb2c3d5832e0fa692/Yank/yank.py#L806-L809). Otherwise, it creates a new `MultistateSampler`.

**Comment by zhang-ivy** [comment]
@mikemhenry  Here is a pickled hybrid topology factory for a cdk2 transformation as well as two sample scripts -- one for running h-repex and one for resuming. You'll just need to change some of the input file names but otherwise they should work. Let me know if you have any problems
[small_molecule_repex.zip](https://github.com/choderalab/perses/files/6804235/small_molecule_repex.zip)

**Comment by jchodera** [comment]
@zhang-ivy @mikemhenry : Our focus here should be on the simplest/fastest example that runs so we can incorporate it into unit tests: Something like a mutation for alanine dipeptide in vacuum would be optimal here. Could you help set up an example like this, @zhang-ivy ?

**Comment by zhang-ivy** [comment]
During our perses dev sync I think we had agreed on using a small molecule transformation for this test case because we wanted to match Relay's applications as closely as possible. Alanine dipeptide in vacuum should be fine too -- I think Mike already has a pickled htf alanine dipeptide in vacuum from debugging the repex cycling issue, but I can re-send a pickled htf tomorrow if you need @mikemhenry

**Comment by mikemhenry** [comment]
@zhang-ivy can you send me a small molecule example that starts "from scratch"? The issue with pickles for this is as we change things in the package, stuff breaks:
``` python
Traceback (most recent call last):
  File "run_h_repex.py", line 10, in <module>
    htf = pickle.load(open("cdk2_solvent.pickle", "rb"))
ModuleNotFoundError: No module named 'oechem'
```

This doesn't work anymore since newer versions of `perses` don't put `oechem` in the package namespace. I can fix this by adding a `from openeye import oechem` to the top of the `run_h_repex.py` script, but for turning this into a unit test we will want something more robust.

**Comment by jchodera** [comment]
@zhang-ivy : You recently did some capped peptide transformations to validate the new reaction field PR. These would be great to include as examples/tests, especially if we can do it without solvent present or with very minimal solvent (e.g. low density)!

**Comment by zhang-ivy** [comment]
I’m on the way to the airport now and will be on vacation until 7/25, but I’ve slack messaged Mike the code to generate a htf for alanine dipeptide in vacuum

**Comment by mikemhenry** [comment]
I've got everything I need and so far resuming is working fine! Just working on the unit tests now.

**Comment by dominicrufa** [comment]
I'm not sure you would want to resume a simulation from a yaml (unless you want to change some simulation parameters explicitly) . in replica-exchange mode, for example, the reporter is written to `.nc` files. the files can be loaded back into `perses.samplers.multistate.HybridRepexSampler` and `openmmtools.multistate.MultiStateReporter` and extended like (or some variation of):

```
from perses.samplers.multistate import HybridRepexSampler
from openmmtools.multistate import MultistateReporter

reporter = MultiStateReporter('my_nc.nc')
simulation = HybridRepexSampler.from_storage(reporter)

total_steps = {N}
run_so_far = simulation.iteration
left_to_do = total_steps - run_so_far

simulation.extend(n_iterations = left_to_do)
```

**Comment by mikemhenry** [comment]
@dominicrufa @jchodera 
I was thinking about a situation where the simulation crashed or you just wanted to run it longer. If we don't have a need to support that and want to just resume with the API, then things are working as intended. 

Now that we have resuming tested, I think we are good for 0.9.1 release and if we want to add functionality where re-running the same yaml will automatically check for a checkpoint file and just resume if it finds one (and makes sure nothing else changed) then that should go in the 0.10.0 milestone.

**Comment by ijpulidos** [comment]
We need more detailed tests to ensure simulations can be resumed. Such as, depending on the ensemble, we could check the sanity of some quantities (positions, velocities, energy, etc.) right after resuming to actually ensure we are doing it correctly. @mikemhenry did a great job with https://github.com/choderalab/perses/pull/821 to ensure this at some level, but I believe we might need more detailed tests.

**Comment by mikemhenry** [comment]
I agree @ijpulidos we also need to test to see if `SIGKILL` will be an issue or not. Now that we have basic functionality tested, I'm going to move this into the next milestone.

**Comment by mikemhenry** [comment]
Fixed with #830
